trinity/rpc/modules/debug.py

```diff
@@ -0,0 +1,158 @@
+from mypy_extensions import (
+    TypedDict,
+)
+
+from typing import (
+    AnyStr,
+    cast,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    TYPE_CHECKING,
+    Union,
+)
+
+from eth_typing import (
+    Hash32,
+)
+
+from eth_utils import (
+    decode_hex,
+    encode_hex,
+    remove_0x_prefix,
+)
+
+from eth_utils.types import (
+    is_string,
+)
+
+from eth.utils.padding import (
+    pad32,
+)
+
+from eth.vm.tracing import (
+    BaseTracer,
+    StructLogEntry,
+    StructTracer,
+)
+
+from trinity.rpc.modules import (
+    RPCModule,
+)
+
+if TYPE_CHECKING:
+    from trinity.chains.base import BaseAsyncChain  # noqa: F401
+
+
+class TraceConfigRPC(TypedDict):
+    debug: Optional[bool]
+    disableMemory: Optional[bool]
+    disableStack: Optional[bool]
+    disableStorage: Optional[bool]
+    limit: Optional[int]
+
+
+class StructLogEntryRPC(TypedDict, total=False):
+    depth: int
+    error: Optional[str]
+    gas: int
+    gasCost: int
+    memory: List[str]
+    op: str
+    pc: int
+    stack: List[str]
+    storage: Dict[int, str]
+
+
+class ExecutionResultRPC(TypedDict):
+    gas: int
+    failed: bool
+    returnValue: str
+    structLogs: List[StructLogEntryRPC]
+
+
+def format_struct_logs(logs: Tuple[StructLogEntry, ...]) -> List[StructLogEntryRPC]:
+    def to_zeropad32_str(value: Union[AnyStr, int]) -> str:
+        val: Union[AnyStr, int] = value
+        if isinstance(value, int):
+            hexstr = '{0:x}'.format(value)
+            if len(hexstr) % 2 != 0:
+                hexstr = '0' + hexstr  # avoid "Odd-length string" error
+            val = bytes.fromhex(hexstr)  # type: ignore
+        elif not is_string(value):
+            raise TypeError("Value must be an instance of str or unicode")
+        return remove_0x_prefix(encode_hex(pad32(val)))
+
+    logs_formatted = []
+    for e in logs:
+        entry = StructLogEntryRPC(depth=e.depth,
+                                  gas=e.gas,
+                                  gasCost=e.gas_cost,
+                                  memory=[],
+                                  op=e.op,
+                                  pc=e.pc,
+                                  stack=[],
+                                  storage={})
+        if e.err:
+            entry['error'] = str(e.err)
+
+        i = 0
+        while i + 32 <= len(e.memory):
+            entry['memory'].append(remove_0x_prefix(encode_hex(e.memory[i: i + 32])))
+            i += 32
```

Comment:
This is a section I'd like to cleanup but it doesn't have to be in this PR.  I'm actually :+1: on removing all of the JSON-RPC stuff from this PR and doing it in a subsequent one (and just merging the core tracer API in this PR.